### PR TITLE
Chapter 12

### PR DESCRIPTION
**Pull request by Moonforeva**

### CODE REVIEW

**Comment by Moonforeva** [comment]
Ready

File: app/controllers/password_resets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: %i(edit update)
+  before_action :valid_user, only: %i(edit update)
+  before_action :check_expiration, only: %i(edit update)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:4]:
```suggestion
  before_action :get_user, :valid_user, only, :check_expiration, only: %i(edit update)
```

File: app/controllers/password_resets_controller.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  def new; end
+
+  def create
+    @user = User.find_by(email: params[:password_reset][:email].downcase)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:9]:
```suggestion
    @user = User.find_by email: params[:password_reset][:email].downcase
```

File: app/controllers/password_resets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    if @user
+      @user.create_reset_digest
+      @user.send_password_reset_email
+      flash[:info] = "Email sent with password reset instructions"
+      redirect_to root_url
+    else
+      flash.now[:danger] = "Email address not found"
+      render 'new'
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:17]:
```suggestion
      render :new
```

File: app/controllers/password_resets_controller.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty?
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:24]:
```suggestion
    if params[:user][:password].blank?
```

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    end
+  end
+
+  def edit; end
+
+  def update
+    if params[:user][:password].empty? 
+      @user.errors.add(:password, "can't be empty")
+      render :edit
+    elsif @user.update(user_params)
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:27]:
```suggestion
    elsif @user.update user_params 
```

File: app/controllers/password_resets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  def update
+    if params[:user][:password].empty? 
+      @user.errors.add(:password, "can't be empty")
+      render :edit
+    elsif @user.update(user_params) 
+      log_in @user
+      flash[:success] = "Password has been reset."
+      redirect_to @user
+    else
+      render :edit
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:32]: add flash for this case

File: app/helpers/password_resets_helper.rb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+module PasswordResetsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/password_resets_helper.rb:2]: remove file that never used

File: app/models/user.rb [hunk 8]
```diff
@@ -52,6 +52,20 @@ def activate
   def send_activation_email
     UserMailer.account_activation(self).deliver_now
   end
+
+  def create_reset_digest
+    self.reset_token = User.new_token
+    update_attribute(:reset_digest, User.digest(reset_token))
+    update_attribute(:reset_sent_at, Time.zone.now)
```
Comment (saobv-2223) [inline app/models/user.rb:59]:
```suggestion
    update reset_digest: User.digest(reset_token), reset_sent_at: Time.zone.now
```

File: app/models/user.rb [hunk 9]
```diff
@@ -52,6 +52,20 @@ def activate
...
+  end
+    
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end
+
+  def password_reset_expired?
+    reset_sent_at < 2.hours.ago
```
Comment (saobv-2223) [inline app/models/user.rb:67]: move 2 to setting.yml file and call here

File: app/views/password_resets/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,20 @@
+<% provide :title, 'Reset password' %>
+<h1>Reset password</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: password_reset_path(params[:id])) do |f| %>
```
Comment (saobv-2223) [inline app/views/password_resets/edit.html.erb:6]: remove ()

File: app/views/password_resets/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+      <%= render 'shared/error_messages' %>
```
Comment (saobv-2223) [inline app/views/password_resets/edit.html.erb:7]: using "

File: app/views/password_resets/edit.html.erb [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+      <%= hidden_field_tag :email, @user.email %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: 'form-control' %>
+
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation, class: 'form-control' %>
```
Comment (saobv-2223) [inline app/views/password_resets/edit.html.erb:15]: using "

File: app/views/password_resets/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+<% provide :title, "Forgot password" %>
+<h1>Forgot password</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:password_reset, url: password_resets_path) do |f| %>
```
Comment (saobv-2223) [inline app/views/password_resets/new.html.erb:6]: remove ()

File: config/routes.rb [hunk 14]
```diff
@@ -10,5 +10,6 @@
   post "/login", to: "sessions#create"
   delete "/logout", to: "sessions#destroy"
   resources :users
-  resources :account_activations, only: :edit
+  resources :account_activations, only: %i(edit)
+  resources :password_resets, only: [:new, :create, :edit, :update]
```
Comment (saobv-2223) [inline config/routes.rb:14]: image

**Comment by Moonforeva** [comment]
Ready

File: app/views/user_mailer/password_reset.html.erb [hunk 15]
```diff
@@ -1,5 +1,9 @@
...
-
+<h1>Password reset</h1>
+<p>To reset your password click the link below:</p>
+<%= link_to "Reset password", edit_password_reset_url(@user.reset_token,
+email: @user.email) %>
+<p>This link will expire in two hours.</p>
 <p>
-  <%= @greeting %>, find me in app/views/user_mailer/password_reset.html.erb
+If you did not request your password to be reset, please ignore this email and
+your password will stay as it is.
```
Comment (saobv-2223) [inline app/views/user_mailer/password_reset.html.erb:8]:
```suggestion
  If you did not request your password to be reset, please ignore this email and
  your password will stay as it is.
```

File: app/controllers/password_resets_controller.rb [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def get_user
+    @user = User.find_by email: params[:email]
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:42]:
```suggestion
    @user = User.find_by email: params[:email]
    
    return if @user
    
    flash[:danger] = "Message"
    redirect_to some_path
```

**Comment by Moonforeva** [comment]
Ready

File: app/controllers/password_resets_controller.rb [hunk 17]
```diff
@@ -0,0 +1,63 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, :valid_user, only, :check_expiration, only: %i(edit update)
```
Comment (manhbnt-0960) [inline app/controllers/password_resets_controller.rb:2]:
```suggestion
  before_action :get_user, :valid_user, :check_expiration, only: %i(edit update)
```

File: app/controllers/password_resets_controller.rb [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    @user = User.find_by email: params[:email]
+
+    return if @user
+    
+    flash[:danger] = "User not found"
+    redirect_to root_url
+  end
+
+  def valid_user
+    unless @user && @user.activated? &&
```
Comment (manhbnt-0960) [inline app/controllers/password_resets_controller.rb:51]:
```suggestion
    return if @user && @user.activated? &&
```

File: app/controllers/password_resets_controller.rb [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+      @user.authenticated? :reset, params[:id]
+      redirect_to root_url
+    end
```
Comment (manhbnt-0960) [inline app/controllers/password_resets_controller.rb:54]: remove "end" keyword

File: app/controllers/password_resets_controller.rb [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  def check_expiration
+    if @user.password_reset_expired?
```
Comment (manhbnt-0960) [inline app/controllers/password_resets_controller.rb:58]:
```suggestion
    return unless @user.password_reset_expired?
```

**Comment by Moonforeva** [comment]
Ready

File: app/controllers/password_resets_controller.rb [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+
+    return if @user
+
+    flash[:danger] = "User not found"
+    redirect_to root_url
+  end
+
+  def valid_user
+    return if @user && @user.activated? &&
+    @user.authenticated? :reset, params[:id]
```
Comment (saobv-2223) [inline app/controllers/password_resets_controller.rb:52]:
```suggestion
    return if @user && @user.activated? && @user.authenticated? :reset, params[:id]
    
```

**Comment by Moonforeva** [comment]
Ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**